rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+                    if input != MAGIC {
+                        self.failed = true;
+                        return Err(McapError::BadMagic);
+                    }
+                    return Ok(None);
+                }
+                Record => {
+                    // need one byte for opcode, then 8 bytes for record length.
+                    let span = match self.load(9)? {
+                        Span(span) => span,
+                        Remainder(want) => return self.request(want),
+                    };
+                    let input = self.uncompressed.view(span);
+                    let opcode = input[0];
+                    let record_length: u64 = u64::from_le_bytes(input[1..].try_into().unwrap());
+                    if opcode == op::CHUNK && !self.options.emit_chunks {
+                        match self.consume(9)? {
+                            Span(_) => {}
+                            Remainder(_) => panic!("there should be 9 bytes available"),
+                        };
+                        self.currently_reading =
+                            CurrentlyReading::ChunkHeader { len: record_length };
+                        continue;
+                    }
+                    // get the rest of the record now.
+                    let span = match self.consume(9 + len_as_usize(record_length)?)? {
+                        Span(span) => span,
+                        Remainder(want) => return self.request(want),
+                    };
+                    // some opcodes trigger state changes.
+                    match opcode {
+                        // A footer implies no more records in the MCAP.
+                        op::FOOTER => self.currently_reading = EndMagic,
+                        // Data end implies end of data section - validate the CRC if present.
+                        op::DATA_END => {
+                            if let Some(calculated) = self.calculated_data_section_crc {
+                                let record_data = &self.uncompressed.view(span)[9..];
+                                match parse_record(opcode, record_data)? {
+                                    crate::records::Record::DataEnd(end) => {
+                                        if end.data_section_crc != 0
+                                            && end.data_section_crc != calculated
+                                        {
+                                            return Err(McapError::BadDataCrc {
+                                                saved: end.data_section_crc,
+                                                calculated,
+                                            });
+                                        }
+                                    }
+                                    _ => unreachable!("should not recieve any other record type"),
+                                }
+                            }
+                            self.data_section_hasher = None;
```

Comment:
why bother clearing this?

Reply:
By clearing this we signal to ourselves that we're not in the data section anymore, and we do not need to keep updating a CRC value every time more file data is read in.